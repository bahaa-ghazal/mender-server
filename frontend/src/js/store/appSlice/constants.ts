// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import { BarChart as BarChartIcon, PieChartOutline as PieChartIcon } from '@mui/icons-material';

import FlagCN from '../../../assets/img/flag-cn.svg';
import FlagEU from '../../../assets/img/flag-eu.svg';
import FlagUS from '../../../assets/img/flag-us.svg';

const startingDeviceCount = {
  os: 'for first 50 devices',
  professional: 'for first 250 devices'
};

export const apiRoot = '/api/management';
export const apiUrl = {
  v1: `${apiRoot}/v1`,
  v2: `${apiRoot}/v2`
};

export const headerNames = {
  link: 'link',
  location: 'location',
  total: 'x-total-count'
};

export const chartTypes = {
  bar: { key: 'bar', Icon: BarChartIcon },
  pie: { key: 'pie', Icon: PieChartIcon }
};
export const emptyChartSelection = { software: '', group: '', chartType: chartTypes.bar.key, attribute: 'artifact_name' };
export const defaultReportType = 'distribution';
export const defaultReports = [{ ...emptyChartSelection, group: null, attribute: 'artifact_name', type: defaultReportType }];

export const BEGINNING_OF_TIME = '2016-01-01T00:00:00.000Z';

export const locations = {
  cn: { key: 'cn', title: 'China', location: 'cn.hosted.mender.io', icon: FlagCN },
  eu: { key: 'eu', title: 'EU', location: 'eu.hosted.mender.io', icon: FlagEU },
  us: { key: 'us', title: 'US', location: 'hosted.mender.io', icon: FlagUS }
};
export type AvailablePlans = 'os' | 'professional' | 'enterprise';

export type Plan = {
  description: string;
  deviceCount: string;
  features: string[];
  id: AvailablePlans;
  minimalDeviceCount: number;
  name: string;
  offer?: boolean;
  offerprice?: string;
  price: string;
  price2?: string;
};

export type AvailableAddon = 'configure' | 'troubleshoot' | 'monitor';

export const PLANS: { [key in AvailablePlans]: Plan } = {
  os: {
    id: 'os',
    name: 'Basic',
    minimalDeviceCount: 50,
    offer: true,
    price: '$34 / month',
    deviceCount: startingDeviceCount.os,
    description:
      'The core features of Mender. To continue using Enterprise Trial features—like Delta updates, scheduled deployments, phased rollouts, device filtering, dynamic groups, RBAC, audit logs, and more—please upgrade to a higher plan.',
    features: ['Access to core features of Mender', 'Basic support']
  },
  professional: {
    id: 'professional',
    name: 'Professional',
    minimalDeviceCount: 250,
    offer: true,
    price: '$291 / month',
    deviceCount: startingDeviceCount.professional,
    description: 'Everything in Basic, plus enhanced update management and automation features.',
    features: ['Advanced OTA features', 'Higher priority support']
  },
  enterprise: {
    id: 'enterprise',
    name: 'Enterprise',
    minimalDeviceCount: 1000,
    price: 'Custom pricing',
    deviceCount: 'Unlimited devices',
    description:
      'Every advanced feature of Mender, tailored for complex and large-scale deployments. Not available as a monthly subscription — ask us for a quote.',
    features: ['All Mender features', 'Advanced security features', 'SLA-backed support']
  }
} as const;
export type Addon = {
  description: string;
  eligible: AvailablePlans[];
  id: AvailableAddon;
  needs: string[];
  title: string;
} & {
  [key in Exclude<AvailablePlans, 'enterprise'>]: {
    deviceCount: string;
    price: string;
  };
};

export type AddonId = keyof typeof ADDONS;

// the needs names need to be aligned with the name of the features in the appReducer, as they will be checked in the addonselection
export const ADDONS: { [key in AvailableAddon]: Addon } = {
  configure: {
    id: 'configure',
    title: 'Configure',
    description: 'Seamlessly configure applications and devices remotely – configure each device to its environment.',
    needs: ['hasDeviceConfig'],
    os: {
      price: '$11/month',
      deviceCount: startingDeviceCount.os
    },
    professional: {
      price: '$70/month',
      deviceCount: startingDeviceCount.professional
    },
    eligible: ['os', 'professional', 'enterprise']
  },
  monitor: {
    id: 'monitor',
    title: 'Monitor',
    description: 'Secure, remote access to your devices – quickly diagnose and fix issues in real time.',
    needs: ['hasMonitor'],
    os: {
      price: '-',
      deviceCount: '-'
    },
    professional: {
      price: '$93/month',
      deviceCount: startingDeviceCount.professional
    },
    eligible: ['professional', 'enterprise']
  },
  troubleshoot: {
    id: 'troubleshoot',
    title: 'Troubleshoot',
    description: 'Detect and analyze health issues of devices, services and applications. Set up alerts so you can act quickly.',
    needs: ['hasDeviceConnect'],
    os: {
      price: '$28/month',
      deviceCount: startingDeviceCount.os
    },
    professional: {
      price: '$79/month',
      deviceCount: startingDeviceCount.professional
    },
    eligible: ['os', 'professional', 'enterprise']
  }
};

export const BENEFITS = {
  auditlog: { id: 'auditlog', benefit: 'trace change across your devices and access troubleshooting session replay', requiredPlan: PLANS.professional.id },
  dashboard: { id: 'dashboard', benefit: 'actionable insights into the devices you are updating with Mender', requiredPlan: PLANS.enterprise.id },
  deltaGeneration: {
    id: 'deltaGeneration',
    benefit: 'automatic delta artifacts generation to minimize data transfer and improve the update delivery',
    requiredPlan: PLANS.enterprise.id
  },
  deviceConfiguration: { id: 'deviceConfiguration', benefit: 'device configuration features', requiredAddon: ADDONS.configure.id },
  deviceMonitor: { id: 'deviceMonitor', benefit: 'device monitoring features', requiredAddon: ADDONS.monitor.id, requiredPlan: PLANS.professional.id },
  deviceTroubleshoot: { id: 'deviceTroubleshoot', benefit: 'device troubleshooting features', requiredAddon: ADDONS.troubleshoot.id },
  dynamicGroups: { id: 'dynamicGroups', benefit: 'create dynamic groups to ease device management', requiredPlan: PLANS.enterprise.id },
  fullFiltering: { id: 'fullFiltering', benefit: 'filtering by multiple attributes to improve the device overview', requiredPlan: PLANS.professional.id },
  gateway: { id: 'gateway', benefit: 'see devices connected to your gateway device for easy access', requiredPlans: PLANS.professional.id },
  pausedDeployments: {
    id: 'pausedDeployments',
    benefit: 'granular control about update rollout to allow synchronization across your fleet',
    requiredPlan: PLANS.enterprise.id
  },
  phasedDeployments: { id: 'phasedDeployments', benefit: 'choose to roll out deployments in multiple phases', requiredPlan: PLANS.enterprise.id },
  rbac: { id: 'rbac', benefit: 'granular role based access control', requiredPlan: PLANS.enterprise.id },
  retryDeployments: { id: 'retryDeployments', benefit: 'optional retries for failed rollout attempts', requiredPlan: PLANS.professional.id },
  scheduledDeployments: {
    id: 'scheduledDeployments',
    benefit: 'scheduled deployments to steer the distribution of your updates.',
    requiredPlan: PLANS.professional.id
  },
  webhookEvents: {
    id: 'webhookEvents',
    benefit: 'receive inventory events and select which type(s) of events the webhook will receive',
    requiredPlan: PLANS.professional.id
  },

  default: { id: 'default', benefit: 'gain access to this feature', requiredPlan: PLANS.enterprise.id }
};

export const yes = () => true;
export const canAccess = yes;

export const DARK_MODE = 'dark';
export const LIGHT_MODE = 'light';

export const APPLICATION_JSON_CONTENT_TYPE = 'application/json';
export const APPLICATION_JWT_CONTENT_TYPE = 'application/jwt';
