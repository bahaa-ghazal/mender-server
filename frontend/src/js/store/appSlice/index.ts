// Copyright 2023 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
// @ts-nocheck
import type { SnackbarProps } from '@mui/material';

import { SORTING_OPTIONS } from '@northern.tech/store/constants';
import { createSlice } from '@reduxjs/toolkit';

import { SortOptions } from '../organizationSlice/types';

export const sliceName = 'app';

const getYesterday = () => {
  const today = new Date();
  today.setDate(today.getDate() - 1);
  return today.toISOString();
};

interface SnackbarContent extends Pick<SnackbarProps, 'action' | 'autoHideDuration' | 'message' | 'open'> {
  preventClickToCopy?: boolean;
}

interface SearchState {
  deviceIds: string[];
  searchTerm: string;
  searchTotal: number;
  sort: SortOptions;
}

interface Upload {
  cancelSource: any;
  progress: number;
}

export type SentryConfig = {
  isReduxEnabled?: string; // no parsing done here as this has to be accessed outside of the store
  location: string;
  replaysSessionSampleRate?: number;
  tracesSampleRate?: number;
};

type AppSliceType = {
  cancelSource: any;
  commit: string;
  demoArtifactLink: string;
  docsVersion: string;
  features: Record<string, boolean>;
  feedbackProbability: number;
  firstLoginAfterSignup: boolean;
  hostAddress: string | null;
  hostedAnnouncement: string;
  newThreshold: string;
  offlineThreshold: string;
  recaptchaSiteKey: string;
  searchState: SearchState;
  sentry: SentryConfig;
  snackbar: SnackbarContent;
  stripeAPIKey: string;
  trackerCode: string;
  uploadsById: Record<string, Upload>;
  versionInformation: Record<string, string>;
  yesterday?: string;
};

export const initialState: AppSliceType = {
  cancelSource: undefined,
  commit: '',
  demoArtifactLink: 'https://dgsbl4vditpls.cloudfront.net/mender-demo-artifact.mender',
  hostAddress: null,
  snackbar: {
    action: undefined,
    autoHideDuration: undefined,
    message: '',
    open: false,
    preventClickToCopy: false
  },
  // return boolean rather than organization details
  features: {
    hasAuditlogs: false,
    hasDeltaProgress: false,
    hasMultitenancy: false,
    hasDeviceConfig: false,
    hasDeviceConnect: false,
    hasFeedbackEnabled: false,
    hasMonitor: false,
    hasReporting: false,
    isHosted: true,
    isEnterprise: false
  },
  feedbackProbability: 0.3,
  firstLoginAfterSignup: false,
  hostedAnnouncement: '',
  docsVersion: '',
  recaptchaSiteKey: '',
  searchState: {
    deviceIds: [],
    searchTerm: '',
    searchTotal: 0,
    sort: {
      direction: SORTING_OPTIONS.desc
      // key: null,
      // scope: null
    }
  },
  sentry: {
    location: '',
    replaysSessionSampleRate: 0.1,
    tracesSampleRate: 1.0
  },
  stripeAPIKey: '',
  trackerCode: '',
  uploadsById: {
    // id: { progress: 0, cancelSource: undefined }
  },
  newThreshold: getYesterday(),
  offlineThreshold: getYesterday(),
  versionInformation: {
    Integration: '',
    'Mender-Client': '',
    'Mender-Artifact': '',
    'Meta-Mender': ''
  },
  yesterday: undefined
};

export const appSlice = createSlice({
  name: sliceName,
  initialState,
  reducers: {
    setFeatures: (state, action) => {
      state.features = {
        ...state.features,
        ...action.payload
      };
    },
    setSnackbar: (state, { payload }) => {
      if (typeof payload === 'string' || payload instanceof String) {
        state.snackbar = {
          ...initialState.snackbar,
          message: payload,
          open: !!payload
        };
        return;
      }
      const { message, autoHideDuration, action, preventClickToCopy = false } = payload;
      state.snackbar = {
        action,
        autoHideDuration,
        message,
        open: !!message,
        preventClickToCopy
      };
    },
    setFirstLoginAfterSignup: (state, action) => {
      state.firstLoginAfterSignup = action.payload;
    },
    setAnnouncement: (state, action) => {
      state.hostedAnnouncement = action.payload;
    },
    setSearchState: (state, action) => {
      state.searchState = {
        ...state.searchState,
        ...action.payload
      };
    },
    setOfflineThreshold: (state, action) => {
      state.offlineThreshold = action.payload;
    },
    initUpload: (state, action) => {
      const { id, upload } = action.payload;
      state.uploadsById[id] = upload;
    },
    uploadProgress: (state, action) => {
      const { id, progress } = action.payload;
      state.uploadsById[id] = {
        ...state.uploadsById[id],
        progress
      };
    },
    cleanUpUpload: (state, action) => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { [action.payload]: current, ...remainder } = state.uploadsById;
      state.uploadsById = remainder;
    },
    setVersionInformation: (state, action) => {
      state.versionInformation = {
        ...state.versionInformation,
        ...action.payload
      };
    },
    setEnvironmentData: (state, action) => ({ ...state, ...action.payload })
  }
});

export const actions = appSlice.actions;
export default appSlice.reducer;
