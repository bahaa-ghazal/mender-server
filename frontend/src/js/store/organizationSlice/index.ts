// Copyright 2023 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import { DEVICE_LIST_DEFAULTS, SORTING_OPTIONS, TENANT_LIST_DEFAULT } from '@northern.tech/store/constants';
import { OrganizationState } from '@northern.tech/store/organizationSlice/types';
import { createSlice } from '@reduxjs/toolkit';

export const sliceName = 'organization';

export const initialState: OrganizationState = {
  card: {
    last4: '',
    expiration: { month: 1, year: 2020 },
    brand: ''
  },
  intentId: null,
  tenantList: {
    ...TENANT_LIST_DEFAULT,
    total: 0,
    tenants: [],
    selectedTenant: null,
    sort: {
      direction: SORTING_OPTIONS.desc,
      key: 'name'
    }
  },
  organization: {
    // id, name, status, tenant_token, plan
  },
  auditlog: {
    events: [],
    selectionState: {
      ...DEVICE_LIST_DEFAULTS,
      detail: undefined,
      endDate: undefined,
      selectedIssue: undefined,
      sort: { direction: SORTING_OPTIONS.desc },
      startDate: undefined,
      total: 0,
      type: undefined,
      user: undefined
    }
  },
  externalDeviceIntegrations: [
    // { <connection_string|x509|...>, id, provider }
  ],
  ssoConfigs: [],
  webhooks: {
    // [id]: { events: [] }
    // for now:
    events: [],
    eventsTotal: 0
  }
};

export const organizationSlice = createSlice({
  name: sliceName,
  initialState,
  reducers: {
    receiveAuditLogs: (state, action) => {
      const { events, total } = action.payload;
      state.auditlog.events = events;
      state.auditlog.selectionState.total = total;
    },
    setAuditLogState: (state, action) => {
      state.auditlog.selectionState = {
        ...state.auditlog.selectionState,
        ...action.payload
      };
    },
    receiveCurrentCard: (state, action) => {
      state.card = action.payload;
    },
    receiveSetupIntent: (state, action) => {
      state.intentId = action.payload;
    },
    setOrganization: (state, action) => {
      state.organization = { ...state.organization, ...action.payload };
    },
    setSubscription: (state, action) => {
      state.organization.subscription = { ...state.organization.subscription, ...action.payload };
    },
    setBillingProfile: (state, action) => {
      state.organization.billing_profile = action.payload;
    },
    setTenantListState: (state, action) => {
      state.tenantList = action.payload;
    },
    receiveExternalDeviceIntegrations: (state, action) => {
      state.externalDeviceIntegrations = action.payload;
    },
    receiveSsoConfigs: (state, action) => {
      state.ssoConfigs = action.payload;
    },
    receiveWebhookEvents: (state, action) => {
      const { value, total } = action.payload;
      state.webhooks.events = value;
      state.webhooks.eventsTotal = total;
    }
  }
});

export const actions = organizationSlice.actions;
export default organizationSlice.reducer;
