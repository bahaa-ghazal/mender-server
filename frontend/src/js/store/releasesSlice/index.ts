// Copyright 2023 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
// @ts-nocheck
import { DEVICE_LIST_DEFAULTS, SORTING_OPTIONS } from '@northern.tech/store/constants';
import { createSlice } from '@reduxjs/toolkit';

import { Release } from '../api/types/Release';
import { SortOptions } from '../organizationSlice/types';

export const sliceName = 'releases';

// TODO: take the following from the types package once synced
enum DeltaJobDetailsItemStatus {
  PENDING = 'pending',
  QUEUED = 'queued',
  SUCCESS = 'success',
  FAILED = 'failed',
  ARTIFACT_UPLOADED = 'artifact_uploaded'
}

type DeltaJobDetailsItem = {
  delta_artifact_size?: number;
  deployment_id?: string;
  devices_types_compatible?: Array<string>;
  exit_code?: number;
  from_release?: string;
  log?: string;
  status?: DeltaJobDetailsItemStatus;
  to_artifact_size?: number;
  to_release?: string;
};

type DeltaJobsListItem = {
  delta_job_id?: string;
  devices_types_compatible?: Array<string>;
  from_version?: string;
  id?: string;
  started?: string;
  status?: DeltaJobDetailsItemStatus;
  to_version?: string;
};

type EnhancedJobDetailsItem = DeltaJobDetailsItem & DeltaJobsListItem;

type ReleaseSliceType = {
  artifacts: never[];
  byId: Record<string, Release>;
  deltaJobs: Record<string, EnhancedJobDetailsItem>;
  deltaJobsList: {
    jobIds: string[];
    page: number;
    perPage: number;
    sort?: SortOptions;
    total: 0;
  };
  releasesList: {
    isLoading?: boolean;
    page: number;
    perPage: number;
    releaseIds: string[];
    searchedIds: string[];
    searchTerm: string;
    searchTotal: number;
    selectedTags: string[];
    selection: number[];
    sort?: SortOptions;
    tab: 'releases' | 'delta';
    total: number;
    type: string;
  };
  selectedJob: string | null;
  selectedRelease: string | null;
  tags: string[];
  updateTypes: string[];
};

export const initialState: ReleaseSliceType = {
  /*
   * Return list of saved artifacts objects
   */
  /*
   * return list of artifacts where duplicate names are collated with device compatibility lists combined
   */
  // artifacts: AppStore.getCollatedArtifacts(AppStore.getArtifactsRepo()),
  artifacts: [],
  /*
   * Return list of saved release objects
   */
  byId: {
    /*
    [releaseName]: {
      artifacts: [
        {
          id: '',
          name: '',
          description: '',
          device_types_compatible: [],
          ...
          updates: [{
            files: [
              { size: 123, name: '' }
            ],
            type_info: { type: '' }
          }],
          url: '' // optional
        }
      ],
      modified: ''
      device_types_compatible,
      name: '',
      tags: ['something'],
      notes: ''
    }
    */
  },
  deltaJobs: {
    /*
      [jobId]: {
        id: '',
        ...
      }
      */
  },
  deltaJobsList: {
    ...DEVICE_LIST_DEFAULTS,
    jobIds: [],
    total: 0,
    sort: {
      direction: SORTING_OPTIONS.desc,
      key: 'started'
    }
  },
  releasesList: {
    ...DEVICE_LIST_DEFAULTS,
    searchedIds: [],
    releaseIds: [],
    selection: [],
    sort: {
      direction: SORTING_OPTIONS.desc,
      key: 'modified'
    },
    isLoading: undefined,
    searchTerm: '',
    searchTotal: 0,
    selectedTags: [],
    total: 0,
    type: ''
  },
  tags: [],
  updateTypes: [],
  selectedJob: null,
  /*
   * Return single release with corresponding Artifacts
   */
  selectedRelease: null
};

export const releaseSlice = createSlice({
  name: sliceName,
  initialState,
  reducers: {
    receiveRelease: (state, action) => {
      const { name } = action.payload;
      state.byId[name] = action.payload;
    },
    receiveReleases: (state, action) => {
      state.byId = action.payload;
    },
    receiveReleaseTags: (state, action) => {
      state.tags = action.payload;
    },
    receiveReleaseTypes: (state, action) => {
      state.updateTypes = action.payload;
    },
    removeRelease: (state, action) => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { [action.payload]: toBeRemoved, ...byId } = state.byId;
      state.byId = byId;
      state.selectedRelease = action.payload === state.selectedRelease ? null : state.selectedRelease;
    },
    selectedRelease: (state, action) => {
      state.selectedRelease = action.payload;
    },
    setReleaseListState: (state, action) => {
      state.releasesList = action.payload;
    },
    receivedDeltaJobs: (state, action) => {
      const { jobs, total } = action.payload;
      state.deltaJobs = jobs.reduce(
        (accu, job) => {
          accu[job.id] = {
            ...accu[job.id],
            ...job
          };
          return accu;
        },
        { ...state.deltaJobs }
      );
      state.deltaJobsList.jobIds = jobs.map(job => job.id);
      state.deltaJobsList.total = total;
    },
    receivedDeltaJobDetails: (state, action) => {
      const job = action.payload;
      state.deltaJobs[job.id] = {
        ...state.deltaJobs[job.id],
        ...job
      };
    },
    setDeltaJobsListState: (state, action) => {
      state.deltaJobsList = { ...state.deltaJobsList, ...action.payload };
    }
  }
});

export const actions = releaseSlice.actions;
export default releaseSlice.reducer;
