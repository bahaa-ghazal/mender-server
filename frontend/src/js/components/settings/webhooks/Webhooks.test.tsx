// Copyright 2022 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import { defaultState, render } from '@/testUtils';
import { EXTERNAL_PROVIDER, TIMEOUTS } from '@northern.tech/store/constants';
import { undefineds, webhookEvents } from '@northern.tech/testing/mockData';
import { act, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import WebhookConfiguration from './Configuration';
import Webhooks from './Webhooks';

describe('Webhooks Component', () => {
  it('renders correctly', async () => {
    const { baseElement } = render(<Webhooks />);
    const view = baseElement.firstChild.firstChild;
    expect(view).toMatchSnapshot();
    expect(view).toEqual(expect.not.stringMatching(undefineds));
  });

  it('renders correctly with entries ', async () => {
    const preloadedState = {
      ...defaultState,
      organization: {
        ...defaultState.organization,
        externalDeviceIntegrations: [
          {
            id: '1',
            credentials: { [EXTERNAL_PROVIDER.webhook.credentialsType]: { url: 'https://example.com' } },
            provider: EXTERNAL_PROVIDER.webhook.provider
          }
        ],
        webhooks: {
          ...defaultState.organization.webhooks,
          events: webhookEvents
        }
      }
    };
    const { baseElement } = render(<Webhooks />, { preloadedState });
    const view = baseElement.firstChild.firstChild;
    expect(view).toMatchSnapshot();
    expect(view).toEqual(expect.not.stringMatching(undefineds));
  });

  it('works as expected', async () => {
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
    const preloadedState = {
      ...defaultState,
      organization: {
        ...defaultState.organization,
        externalDeviceIntegrations: [
          {
            id: '1',
            credentials: { [EXTERNAL_PROVIDER.webhook.credentialsType]: { url: 'https://example.com' } },
            provider: EXTERNAL_PROVIDER.webhook.provider
          }
        ],
        webhooks: {
          ...defaultState.organization.webhooks,
          events: webhookEvents
        }
      }
    };
    const ui = <Webhooks />;
    const { rerender } = render(ui, { preloadedState });
    await user.click(screen.getByText('https://example.com'));
    await waitFor(() => rerender(ui));
    expect(screen.getByText(/webhook details/i)).toBeVisible();
    await user.click(screen.getAllByText(/device status updated/i)[0]);
    await waitFor(() => rerender(ui));
    expect(screen.getByText(/Payload/i)).toBeVisible();
    expect(screen.getByRole('button', { name: /delete webhook/i })).toBeDisabled();
    await user.click(screen.getByText(/back to webhook/i));
    await waitFor(() => rerender(ui));
    await user.click(screen.getByLabelText(/close/i));
    await waitFor(() => expect(screen.queryByText(/webhook details/i)).toBeNull());
  });

  it('can be configured', { timeout: 3 * TIMEOUTS.fiveSeconds }, async () => {
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
    const onSubmit = vi.fn();
    render(<WebhookConfiguration onSubmit={onSubmit} />);
    expect(screen.getByText(/save/i)).not.toBeEnabled();
    await user.type(screen.getByLabelText(/url/i), 'http://foo.bar');
    await waitFor(() => expect(screen.queryByText(/not protected by HTTPS/i)).toBeInTheDocument());
    await user.clear(screen.getByLabelText(/url/i));
    await user.type(screen.getByLabelText(/url/i), 'https://foo.bar');
    expect(screen.queryByText(/not protected by HTTPS/i)).not.toBeInTheDocument();
    await user.type(screen.getByLabelText(/Description/i), 'https://foo.bar');
    await user.type(screen.getByLabelText(/secret/i), 'https://foo.bar');
    expect(screen.getByText(/has to be entered as a hexadecimal/i)).toBeVisible();
    await user.clear(screen.getByLabelText(/secret/i));
    await waitFor(() => expect(screen.getByRole('button', { name: /save/i })).toBeEnabled());
    await act(async () => await user.click(screen.getByRole('button', { name: /save/i })));
    expect(onSubmit).toHaveBeenCalledWith({
      credentials: { http: { secret: '', url: 'https://foo.bar' }, type: 'http' },
      description: 'https://foo.bar',
      id: 'new',
      provider: 'webhook',
      scopes: []
    });
  });
});
