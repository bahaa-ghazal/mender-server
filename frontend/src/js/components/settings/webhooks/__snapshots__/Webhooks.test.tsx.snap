// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Webhooks Component > renders correctly 1`] = `null`;

exports[`Webhooks Component > renders correctly with entries  1`] = `
.emotion-0 {
  display: table;
  width: 100%;
  border-collapse: collapse;
  border-spacing: 0;
}

.emotion-0 caption {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  padding: 16px;
  color: rgba(0, 0, 0, 0.6);
  text-align: left;
  caption-side: bottom;
}

.emotion-1 {
  display: table-header-group;
}

.emotion-1 .columnHeader {
  display: table-cell;
}

.emotion-1 .columnHeader .sortIcon {
  margin-bottom: -6px;
}

.emotion-1 .nonSortable {
  cursor: initial;
}

.emotion-2 {
  color: inherit;
  display: table-row;
  vertical-align: middle;
  outline: 0;
}

.emotion-2.MuiTableRow-hover:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

.emotion-2.Mui-selected {
  background-color: rgba(25, 118, 210, 0.08);
}

.emotion-2.Mui-selected:hover {
  background-color: rgba(25, 118, 210, 0.12);
}

.emotion-3 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  display: table-cell;
  vertical-align: inherit;
  border-bottom: 1px solid rgba(224, 224, 224, 1);
  text-align: left;
  padding: 16px;
  color: rgba(0, 0, 0, 0.87);
  line-height: 1.5rem;
  font-weight: 500;
}

.emotion-6 {
  display: table-row-group;
}

.emotion-8 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  display: table-cell;
  vertical-align: inherit;
  border-bottom: 1px solid rgba(224, 224, 224, 1);
  text-align: left;
  padding: 16px;
  color: rgba(0, 0, 0, 0.87);
}

.emotion-11 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: 1.25rem;
}

<div>
  <h2>
    Webhooks
  </h2>
  <table
    class="MuiTable-root margin-bottom emotion-0"
  >
    <thead
      class="MuiTableHead-root emotion-1"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head emotion-2"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium columnHeader nonSortable emotion-3"
          scope="col"
        >
          URL
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium columnHeader nonSortable emotion-3"
          scope="col"
        >
          Description
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium columnHeader nonSortable emotion-3"
          scope="col"
        >
          Manage
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root emotion-6"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-hover clickable emotion-2"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium relative emotion-8"
        >
          https://example.com
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium relative emotion-8"
        />
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium relative emotion-8"
        >
          <div
            class="bold flexbox center-aligned link-color nowrap "
          >
            View details
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall margin-left-xsmall emotion-11"
              data-testid="ArrowForwardIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
              />
            </svg>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
